mipengine/algorithms/algorithm.py

```diff
@@ -170,7 +175,7 @@ def get_specification(cls) -> AlgorithmSpecification:
         return AlgorithmSpecification.parse_file(file)
 
     @abstractmethod
-    def run(self, data, metadata):
+    def run(self, data: "LocalNodesTable", metadata: Dict[str, Dict[str, str]]):
```

Comment:
The `metadata` type is not correct. The actual type is very complex, it's actually something like
```python
Dict[str, Dict[str, Union[str, Dict[str, str], bool, int, float, None]]]
```
Python's dynamic nature makes working with this kind of objects perfectly reasonable, even when the type is horrible. But I don't think the full type adds any value here.

I suggest to type it as just 
```python
Dict[str, dict]
``` 
which can easily be interpreted as a mapping from variable names to metadata dictionaries.